README.md
# python devtools

[![CI](https://github.com/samuelcolvin/python-devtools/workflows/CI/badge.svg?event=push)](https://github.com/samuelcolvin/python-devtools/actions?query=event%3Apush+branch%3Amain+workflow%3ACI)
[![Coverage](https://codecov.io/gh/samuelcolvin/python-devtools/branch/main/graph/badge.svg)](https://codecov.io/gh/samuelcolvin/python-devtools)
[![pypi](https://img.shields.io/pypi/v/devtools.svg)](https://pypi.python.org/pypi/devtools)
[![versions](https://img.shields.io/pypi/pyversions/devtools.svg)](https://github.com/samuelcolvin/python-devtools)
[![license](https://img.shields.io/github/license/samuelcolvin/python-devtools.svg)](https://github.com/samuelcolvin/python-devtools/blob/main/LICENSE)

**Python's missing debug print command and other development tools.**

For more information, see [documentation](https://python-devtools.helpmanual.io/).

## Install

Just

```bash
pip install devtools
```

If you've got python 3.7+ and `pip` installed, you're good to go.

## Usage

```py
from devtools import debug

whatever = [1, 2, 3]
debug(whatever)
```

Outputs:

```py
test.py:4 <module>:
    whatever: [1, 2, 3] (list)
```


That's only the tip of the iceberg, for example:

```py
import numpy as np

data = {
    'foo': np.array(range(20)),
    'bar': {'apple', 'banana', 'carrot', 'grapefruit'},
    'spam': [{'a': i, 'b': (i for i in range(3))} for i in range(3)],
    'sentence': 'this is just a boring sentence.\n' * 4
}

debug(data)
```

outputs:

![python-devtools demo](https://raw.githubusercontent.com/samuelcolvin/python-devtools/main/demo.py.png)

## Usage without Import

devtools can be used without `from devtools import debug` if you add `debug` into `__builtins__`
in `sitecustomize.py`.

For instructions on adding `debug` to `__builtins__`,
see the [installation docs](https://python-devtools.helpmanual.io/usage/#usage-without-import).


HISTORY.md
## v0.12.1 (2023-08-17)

fix docs release

## v0.12.0 (2023-08-17)

* build docs on CI by @samuelcolvin in #127
* Update usage to reflect the recent addition of the pytest plugin by @tomhamiltonstubber in #128
* support dataclasses with slots by @samuelcolvin in #136
* Make `Pygments` required #137

## v0.11.0 (2023-04-05)

* added support for sqlalchemy2 by @the-vty in #120
* switch to ruff by @samuelcolvin in #124
* support displaying ast types by @samuelcolvin in #125
* Insert assert by @samuelcolvin in #126

## v0.10.0 (2022-11-28)

* Use secure builtins standard module, instead of the `__builtins__` by @0xsirsaif in #109
* upgrade executing to fix 3.10 by @samuelcolvin in #110
* Fix windows build by @samuelcolvin in #111
* Allow executing dependency to be >1.0.0 by @staticf0x in #115
* more precise timer summary by @banteg in #113
* Python 3.11 by @samuelcolvin in #118

## v0.9.0 (2022-07-26)

* fix format of nested dataclasses, #99 thanks @aliereno
* Moving to `pyproject.toml`, complete type hints and test with mypy, #107
* add `install` command to add `debug` to `__builtins__`, #108

## v0.8.0 (2021-09-29)

* test with python 3.10 #91
* display `SQLAlchemy` objects nicely #94
* fix tests on windows #93
* show function `qualname` #95
* cache pygments loading (significant speedup) #96

## v0.7.0 (2021-09-03)

* switch to [`executing`](https://pypi.org/project/executing/) and [`asttokens`](https://pypi.org/project/asttokens/)
  for finding and printing debug arguments, #82, thanks @alexmojaki
* correct changelog links, #76, thanks @Cielquan
* return `debug()` arguments, #87
* display more generators like `map` and `filter`, #88
* display `Counter` and similar dict-like objects properly, #88
* display `dataclasses` properly, #88
* uprev test dependencies, #81, #83, #90

## v0.6.1 (2020-10-22)

compatibility with python 3.8.6

## v0.6.0 (2020-07-29)

* improve `__pretty__` to work better with pydantic classes, #52
* improve the way statement ranges are calculated, #58
* drastically improve import time, #50
* pretty printing for non-standard dicts, #60
* better statement finding for multi-line statements, #61
* colors in windows, #57
* fix `debug(type(dict(...)))`, #62

## v0.5.1 (2019-10-09)

* fix python tag in `setup.cfg`, #46

## v0.5.0 (2019-01-03)

* support `MultiDict`, #34
* support `__pretty__` method, #36

## v0.4.0 (2018-12-29)

* remove use of `warnings`, include in output, #30
* fix rendering errors #31
* better str and bytes wrapping #32
* add `len` everywhere possible, part of #16

## v0.3.0 (2017-10-11)

* allow `async/await` arguments
* fix subscript
* fix weird named tuples eg. `mock > call_args`
* add `timer`

## v0.2.0 (2017-09-14)

* improve output
* numerous bug fixes


devtools/ansi.py
from enum import IntEnum

from .utils import isatty

__all__ = 'sformat', 'sprint'

MYPY = False
if MYPY:
    from typing import Any, Mapping, Union


def strip_ansi(value: str) -> str:
    import re

    return re.sub('\033\\[((?:\\d|;)*)([a-zA-Z])', '', value)


class Style(IntEnum):
    reset = 0

    bold = 1
    not_bold = 22

    dim = 2
    not_dim = 22

    italic = 3
    not_italic = 23

    underline = 4
    not_underline = 24

    blink = 5
    not_blink = 25

    reverse = 7
    not_reverse = 27

    strike_through = 9
    not_strike_through = 29

    # foreground colours
    black = 30
    red = 31
    green = 32
    yellow = 33
    blue = 34
    magenta = 35
    cyan = 36
    white = 37

    # background colours
    bg_black = 40
    bg_red = 41
    bg_green = 42
    bg_yellow = 43
    bg_blue = 44
    bg_magenta = 45
    bg_cyan = 46
    bg_white = 47

    # this is a meta value used for the "Style" instance which is the "style" function
    function = -1

    def __call__(self, input: 'Any', *styles: 'Union[Style, int, str]', reset: bool = True, apply: bool = True) -> str:
        """
        Styles text with ANSI styles and returns the new string.

        By default the styling is cleared at the end of the string, this can be prevented with``reset=False``.

        Examples::

            print(sformat('Hello World!', sformat.green))
            print(sformat('ATTENTION!', sformat.bg_magenta))
            print(sformat('Some things', sformat.reverse, sformat.bold))

        :param input: the object to style with ansi codes.
        :param *styles: zero or more styles to apply to the text, should be either style instances or strings
                        matching style names.
        :param reset: if False the ansi reset code is not appended to the end of the string
        :param: apply: if False no ansi codes are applied
        """
        text = str(input)
        if not apply:
            return text
        codes = []
        for s in styles:
            # raw ints are allowed
            if not isinstance(s, self.__class__) and not isinstance(s, int):
                try:
                    s = self.styles[s]
                except KeyError:
                    raise ValueError(f'invalid style "{s}"')
            codes.append(_style_as_int(s.value))  # type: ignore

        if codes:
            r = _as_ansi(';'.join(codes)) + text
        else:
            r = text

        if reset:
            r += _as_ansi(_style_as_int(self.reset))
        return r

    @property
    def styles(self) -> 'Mapping[str, Style]':
        return self.__class__.__members__

    def __repr__(self) -> str:
        if self == self.function:
            return '<pseudo function sformat(text, *styles)>'
        else:
            return super().__repr__()

    def __str__(self) -> str:
        if self == self.function:
            return repr(self)
        else:
            # this matches `super().__str__()` in python 3.7 - 3.10
            # required since IntEnum.__str__ was changed in 3.11,
            # see https://docs.python.org/3/library/enum.html#enum.IntEnum
            return f'{self.__class__.__name__}.{self._name_}'


def _style_as_int(v: 'Union[Style, int]') -> str:
    if isinstance(v, Style):
        return str(v.value)
    else:
        return str(v)


def _as_ansi(s: str) -> str:
    return f'\033[{s}m'


sformat = Style(-1)


class StylePrint:
    """
    Annoyingly enums do not allow inheritance, a lazy design mistake, this is an ugly work around
    for that mistake.
    """

    def __call__(
        self,
        input: str,
        *styles: 'Union[Style, int, str]',
        reset: bool = True,
        flush: bool = True,
        file: 'Any' = None,
        **print_kwargs: 'Any',
    ) -> None:
        text = sformat(input, *styles, reset=reset, apply=isatty(file))
        print(text, flush=flush, file=file, **print_kwargs)

    def __getattr__(self, item: str) -> str:
        return getattr(sformat, item)

    def __repr__(self) -> str:
        return '<pseudo function sprint(text, *styles)>'


sprint = StylePrint()
